inloop/solutions/views.py

```diff
@@ -29,54 +29,111 @@ def get(self, request, id):
         return JsonResponse({'solution_id': solution.id, 'status': solution.status()})
 
 
-class SolutionEditorView(LoginRequiredMixin, View):
+class SolutionSubmissionView(LoginRequiredMixin, View):
+
+    class Error(ValueError):
+        pass
+
+    def get_task(self, slug):
+        try:
+            task = Task.objects.published().get(slug=slug)
+        except Task.DoesNotExist:
+            raise SolutionSubmissionView.Error(
+                'This task does not exist.'
+            )
+        if task.is_expired:
+            raise SolutionSubmissionView.Error(
+                'The deadline for this task has passed.'
+            )
+
+        return task
+
+    def submit(self, files, author, task):
+        if not files:
+            raise SolutionSubmissionView.Error(
+                "You haven't uploaded any files."
+            )
+
+        try:
+            validate_filenames([f.name for f in files])
+        except ValidationError as e:
+            raise SolutionSubmissionView.Error(
+                str(e)
+            )
+
+        with transaction.atomic():
+            solution = Solution.objects.create(author=author, task=task)
+            SolutionFile.objects.bulk_create([
+                SolutionFile(solution=solution, file=f) for f in files
+            ])
+
+        solution_submitted.send(sender=self.__class__, solution=solution)
+
+
+class SolutionEditorView(SolutionSubmissionView):
+
+    class JsonFailureResponse(JsonResponse):
+        def __init__(self):
+            super().__init__({'success': False})
+
+    class JsonSuccessResponse(JsonResponse):
+        def __init__(self):
+            super().__init__({'success': True})
+
     def get(self, request, slug):
         return TemplateResponse(request, 'solutions/editor/editor.html', {
             'task': get_object_or_404(Task.objects.published(), slug=slug),
             'active_tab': 1
         })
 
     def post(self, request, slug):
-        task = get_object_or_404(Task.objects.published(), slug=slug)
```

Comment:
Behavior changed from returning a `404` to returning a `200` with a JSON error message (so technically, this isn't a refactoring).